align/pnr/placer_pythonic.py

```diff
@@ -0,0 +1,116 @@
+import logging
+
+from align.schema.hacks import VerilogJsonTop
+from align.pnr.checker import check_placement
+from align.pnr.placer_pythonic_sp import place_using_sequence_pairs
+
+
+def compute_topoorder(modules, concrete_name, key='abstract_template_name'):
+    found_modules, found_leaves = set(), set()
+    order = list()
+
+    def aux(cn):
+        if cn in modules:
+            found_modules.add(cn)
+            for instance in modules[cn]['instances']:
+                ctn = instance[key]
+                if ctn not in found_modules and ctn not in found_leaves:
+                    aux(ctn)
+            order.append(cn)
+        else:
+            found_leaves.add(cn)
+            order.append(cn)
+    aux(concrete_name)
+
+    return order, found_modules, found_leaves
+
+
+def trim_placement_data(placement_data, top_level):
+
+    modules = {module['concrete_name']: module for module in placement_data['modules']}
+
+    top_concrete_names = [module['concrete_name'] for module in placement_data['modules'] if module['abstract_name'] == top_level]
+    all_modules_leaves = list()
+    for concrete_name in top_concrete_names:
+        _, found_modules, found_leaves = compute_topoorder(modules, concrete_name, key='concrete_template_name')
+        all_modules_leaves.extend(found_modules)
+        all_modules_leaves.extend(found_leaves)
+
+    all_modules_leaves = set(all_modules_leaves)
+
+    new_placement_data = {'leaves': list(), 'modules': list()}
+    new_placement_data['leaves'] = [leaf for leaf in placement_data['leaves'] if leaf['concrete_name'] in all_modules_leaves]
+    new_placement_data['modules'] = [leaf for leaf in placement_data['modules'] if leaf['concrete_name'] in all_modules_leaves]
+
+    for key in ['leaves', 'modules']:
+        for elem in new_placement_data[key]:
+            if 'pin_bbox' in elem:
+                del elem['pin_bbox']
```

Comment:
This will probably remove 'pin_bbox' from the original 'placement_data' because you didn't make a copy of leaf. Is that what you want?

Reply:
It is okay to get rid of `pin_data` as it is not part of the output.